hudi-client/hudi-client-common/src/main/java/org/apache/hudi/execution/CopyOnWriteInsertHandler.java

```diff
@@ -85,7 +85,9 @@ public void consumeOneRecord(HoodieInsertValueGenResult<HoodieRecord> payload) {
       handles.put(partitionPath, handle);
     }
 
-    if (!handle.canWrite(payload.record)) {
+    // If index requires a customized partitioner, cannot open a new file here since it might violate
+    // the distribution required by index.
+    if (!handle.canWrite(payload.record) && config.getNumBuckets() <= 0) {
```

Comment:
I am confused about the `config.getNumBuckets <= 0`, it means requires a customized partitioner?

Reply:
yes, add a new api `HoodieIndex.needCustomizedPartitioner` to do the thing like this